File: R/AoC 2017 Day 13.r
library(tidyverse)



# 

df <-
  tibble(line = read_lines(input)) %>%
  mutate(
    depth = str_extract(line, "^\\d+") %>% parse_integer(),
    range = str_extract(line, "\\d+$") %>% parse_integer(),
    severity = depth * range,
    period = (range - 2) * 2 + 2,
    is_caught = (depth %% period) == 0
  )
df

answer <-
  df %>%
  filter(is_caught) %>%
  pull(severity) %>%
  sum()
answer

# This is too slow
# t <- 0
# repeat {
#   caught <-
#     df %>%
#     mutate(
#       depth = depth + t,
#       period = (range - 2) * 2 + 2,
#       is_caught = (depth %% period) == 0
#     ) %>%
#     pull(is_caught) %>%
#     any()
#   if (!caught) break
#   t <- t + 1
# }

# answer <- t
# answer

Rcpp::cppFunction('
int solve_cpp(std::vector<int> depth, std::vector<int> period) {
  bool done = false;
  int t = 0;
  while (!done) {
    done = true;
    for (int i = 0; i < depth.size(); ++i) {
      if ((depth[i] + t) % period[i] == 0) {
        done = false;
        break;
      }
    }
    ++t;
  }
  return t - 1;
}
')

answer <- solve_cpp(df$depth, df$period)
answer

File: python/AoC 2017 Day 13.py
from aocd import get_data

inp = get_data(day=13, year=2017)

import re

scanners = [[int(x) for x in re.findall(r'\d+', line)] for line in inp.split('\n')]

trip_severity = 0
scanner_properties = []
for depth, r in scanners:
    severity = depth * r
    period = (r - 2) * 2 + 2
    is_caught = (depth % period) == 0

    if is_caught:
        trip_severity += severity
    scanner_properties.append((depth, period))

answer = trip_severity
print(answer)

import itertools

def find_smallest_delay(scanner_properties):
    for t in itertools.count():
        if all((depth + t) % period for depth, period in scanner_properties):
            return t

answer = find_smallest_delay(scanner_properties)
print(answer)
